sweep/sweeper.go

```diff
@@ -1153,43 +1330,58 @@ func (s *UtxoSweeper) CreateSweepTx(inputs []input.Input,
 	return tx, err
 }
 
-// DefaultNextAttemptDeltaFunc is the default calculation for next sweep attempt
-// scheduling. It implements exponential back-off with some randomness. This is
-// to prevent a stuck tx (for example because fee is too low and can't be bumped
-// in btcd) from blocking all other retried inputs in the same tx.
-func DefaultNextAttemptDeltaFunc(attempts int) int32 {
-	return 1 + rand.Int31n(1<<uint(attempts-1))
-}
-
 // ListSweeps returns a list of the sweeps recorded by the sweep store.
 func (s *UtxoSweeper) ListSweeps() ([]chainhash.Hash, error) {
 	return s.cfg.Store.ListSweeps()
 }
 
+// mempoolLookup takes an input's outpoint and queries the mempool to see
+// whether it's already been spent in a transaction found in the mempool.
+// Returns the transaction if found.
+func (s *UtxoSweeper) mempoolLookup(op wire.OutPoint) (*wire.MsgTx, bool) {
+	// For neutrino backend, there's no mempool available, so we exit
+	// early.
+	if s.cfg.Mempool == nil {
+		log.Debugf("Skipping mempool lookup for %v, no mempool ", op)
+
+		return nil, false
+	}
+
+	// Query this input in the mempool. If this outpoint is already spent
+	// in mempool, we should get a spending event back immediately.
+	return s.cfg.Mempool.LookupInputMempoolSpend(op)
+}
+
 // handleNewInput processes a new input by registering spend notification and
 // scheduling sweeping for it.
 func (s *UtxoSweeper) handleNewInput(input *sweepInputMessage) {
 	outpoint := *input.input.OutPoint()
-	pendInput, pending := s.pendingInputs[outpoint]
+	pi, pending := s.pendingInputs[outpoint]
 	if pending {
-		log.Debugf("Already pending input %v received", outpoint)
+		log.Debugf("Already has pending input %v received", outpoint)
 
-		s.handleExistingInput(input, pendInput)
+		s.handleExistingInput(input, pi)
 
 		return
 	}
 
 	// Create a new pendingInput and initialize the listeners slice with
 	// the passed in result channel. If this input is offered for sweep
 	// again, the result channel will be appended to this slice.
-	pendInput = &pendingInput{
-		listeners:        []chan Result{input.resultChan},
-		Input:            input.input,
-		minPublishHeight: s.currentHeight,
-		params:           input.params,
+	pi = &pendingInput{
+		state:     StateInit,
+		listeners: []chan Result{input.resultChan},
+		Input:     input.input,
+		params:    input.params,
 	}
-	s.pendingInputs[outpoint] = pendInput
-	log.Tracef("input %v added to pendingInputs", outpoint)
+
+	// Try to find fee info for possible RBF if this input has already been
+	// spent.
+	pi = s.attachAvailableRBFInfo(pi)
```

Comment:
It seems like there is some ambiguity on the correct way to use this api. afaict `pi =` is not necessary because we mutate `pi` inside this function anyway. In general, as a pattern I'd recommend structuring APIs in a way where we only mutate the receiver. This is more of a general discussion on how we can control mutation semantics so we don't really need to address it here, but this pattern here does sketch me out a bit.